commit 8d8044a2f01a625058db23abbeef76fc2778b983
Author: Pawel Moll <pawel.moll@st.com>
Date:   Wed Aug 6 14:35:32 2008 +0100

    sound_stm: STx7141 initial ALSA driver
    
    Add support for STx7141 audio. LPCM and SPDIF HDMI audio as well as
    internal analogue DACs tested and working.
    There is no support for the external 8 channel DAC on the MB628 in this
    patch.
    
    Signed-off-by: Stephen Gallimore <stephen.gallimore@st.com>
    Signed-off-by: Pawel Moll <pawel.moll@st.com>

diff --git a/arch/sh/kernel/cpu/sh4/setup-stx7141.c b/arch/sh/kernel/cpu/sh4/setup-stx7141.c
index 970ea10..acc5e89 100644
--- a/arch/sh/kernel/cpu/sh4/setup-stx7141.c
+++ b/arch/sh/kernel/cpu/sh4/setup-stx7141.c
@@ -715,6 +715,99 @@ void stx7141_configure_ethernet(int port, int reverse_mii, int mode,
 	platform_device_register(&stx7141eth_devices[port]);
 }
 
+/* Audio output ------------------------------------------------------------ */
+
+void stx7141_configure_audio_pins(int pcmout1, int pcmout2, int spdif,
+		int pcmin1, int pcmin2)
+{
+	/* Claim PIO pins as first PCM player outputs, depending on
+	 * how many DATA outputs are to be used... */
+
+	if (pcmout1 > 0) {
+		stx7141_pio_sysconf(15, 4, 1, "AUDD1_PCMCLKOUT");
+		stpio_request_pin(15, 4, "AUDD1_PCMCLKOUT", STPIO_OUT);
+		stx7141_pio_sysconf(15, 5, 1, "AUDD1_LRCLKOUT");
+		stpio_request_pin(15, 5, "AUDD1_LRCLKOUT", STPIO_OUT);
+		stx7141_pio_sysconf(15, 6, 1, "AUDD1_SCLKOUT");
+		stpio_request_pin(15, 6, "AUDD1_SCLKOUT", STPIO_OUT);
+		stx7141_pio_sysconf(15, 3, 1, "AUDD1_PCMOUT");
+		stpio_request_pin(15, 3, "AUDD1_PCMOUT", STPIO_OUT);
+	}
+	if (pcmout1 > 1) {
+		stx7141_pio_sysconf(15, 7, 2, "AUDD1_PCMOUT[1]");
+		stpio_request_pin(15, 7, "AUDD1_PCMOUT", STPIO_OUT);
+	}
+	if (pcmout1 > 2) {
+		stx7141_pio_sysconf(16, 0, 2, "AUDD1_PCMOUT[2]");
+		stpio_request_pin(16, 0, "AUDD1_PCMOUT", STPIO_OUT);
+	}
+	if (pcmout1 > 3) {
+		stx7141_pio_sysconf(16, 1, 2, "AUDD1_PCMOUT[3]");
+		stpio_request_pin(16, 1, "AUDD1_PCMOUT", STPIO_OUT);
+	}
+	if (pcmout1 > 4) {
+		stx7141_pio_sysconf(16, 2, 2, "AUDD1_PCMOUT[4]");
+		stpio_request_pin(16, 2, "AUDD1_PCMOUT", STPIO_OUT);
+	}
+	if (pcmout1 > 5)
+		BUG();
+
+	/* Claim PIO pins for second PCM player outputs, however
+	 * they are multiplexed with the first player's ones... */
+
+	if (pcmout2 > 0) {
+		if (pcmout1 > 1)
+			BUG();
+
+		stx7141_pio_sysconf(16, 0, 1, "AUDD2_PCMCLKOUT");
+		stpio_request_pin(16, 0, "AUDD2_PCMCLKOUT", STPIO_OUT);
+		stx7141_pio_sysconf(16, 1, 1, "AUDD2_LRCLKOUT");
+		stpio_request_pin(16, 1, "AUDD2_LRCLKOUT", STPIO_OUT);
+		stx7141_pio_sysconf(16, 2, 1, "AUDD2_SCLKOUT");
+		stpio_request_pin(16, 2, "AUDD2_SCLKOUT", STPIO_OUT);
+		stx7141_pio_sysconf(15, 7, 1, "AUDD2_PCMOUT");
+		stpio_request_pin(15, 7, "AUDD2_PCMOUT", STPIO_OUT);
+	}
+	if (pcmout2 > 1)
+		BUG();
+
+	/* Claim PIO pin as SPDIF output... */
+
+	if (spdif > 0) {
+		stx7141_pio_sysconf(16, 3, 1, "AUDD_SPDIFOUT");
+		stpio_request_pin(16, 3, "AUDD_SPDIFOUT", STPIO_OUT);
+	}
+	if (spdif > 1)
+		BUG();
+
+
+	/* Claim PIO for the first PCM reader inputs... */
+
+	if (pcmin1 > 0) {
+		stx7141_pio_sysconf(15, 0, 1, "AUDD1_PCMIN");
+		stpio_request_pin(15, 0, "AUDD1_PCMIN", STPIO_IN);
+		stx7141_pio_sysconf(15, 1, 1, "AUDD1_LRCLKIN");
+		stpio_request_pin(15, 1, "AUDD1_LRCLKIN", STPIO_IN);
+		stx7141_pio_sysconf(15, 2, 1, "AUDD1_SCLKIN");
+		stpio_request_pin(15, 2, "AUDD1_SCLKIN", STPIO_IN);
+	}
+	if (pcmin1 > 1)
+		BUG();
+
+	/* Claim PIO for the second PCM reader inputs... */
+
+	if (pcmin2 > 0) {
+		stx7141_pio_sysconf(16, 4, 1, "AUDD2_PCMIN");
+		stpio_request_pin(16, 4, "AUDD2_PCMIN", STPIO_IN);
+		stx7141_pio_sysconf(16, 5, 1, "AUDD2_LRCLKIN");
+		stpio_request_pin(16, 5, "AUDD2_LRCLKIN", STPIO_IN);
+		stx7141_pio_sysconf(16, 6, 1, "AUDD2_SCLKIN");
+		stpio_request_pin(16, 6, "AUDD2_SCLKIN", STPIO_IN);
+	}
+	if (pcmin2 > 1)
+		BUG();
+}
+
 /* PWM resources ----------------------------------------------------------- */
 
 static struct resource stm_pwm_resource[] = {
diff --git a/include/linux/stm/soc.h b/include/linux/stm/soc.h
index 2cde1c9..1df2544 100644
--- a/include/linux/stm/soc.h
+++ b/include/linux/stm/soc.h
@@ -292,6 +292,8 @@ void stx7141_configure_ssc(struct plat_ssc_data *data);
 void stx7141_configure_usb(int port);
 void stx7141_configure_ethernet(int port, int reverse_mii, int mode,
 				int phy_bus);
+void stx7141_configure_audio_pins(int pcmout1, int pcmout2, int spdif,
+		int pcmin1, int pcmint2);
 
 void stx7200_early_device_init(void);
 void stx7200_configure_asc(const int *ascs, int num_ascs, int console);
diff --git a/sound/stm/Kconfig b/sound/stm/Kconfig
index 1d7cb16..61caa87 100644
--- a/sound/stm/Kconfig
+++ b/sound/stm/Kconfig
@@ -1,7 +1,7 @@
 menuconfig SND_STM
 	tristate "STMicroelectronics SoCs support"
 	depends on SND
-	depends on CPU_SUBTYPE_STB7100 || CPU_SUBTYPE_STX7111 || CPU_SUBTYPE_STX7200
+	depends on CPU_SUBTYPE_STB7100 || CPU_SUBTYPE_STX7111 || CPU_SUBTYPE_STX7141 || CPU_SUBTYPE_STX7200
 	select STM_DMA
 	select SND_PCM
 	select SND_DYNAMIC_MINORS
@@ -76,6 +76,16 @@ config SND_STM_STX7111
 		If you compile this driver as a module, it will be
 		called snd-stm-stx7111.ko
 
+config SND_STM_STX7141
+	tristate "STx7141 support"
+	depends on SND_STM
+	default y if CPU_SUBTYPE_STX7141
+	help
+		Select this to get ALSA driver for STx7141 chip.
+
+		If you compile this driver as a module, it will be
+		called snd-stm-stx7141.ko
+
 config SND_STM_STX7200
 	tristate "STx7200 support"
 	depends on SND_STM
diff --git a/sound/stm/Makefile b/sound/stm/Makefile
index 32e82e5..690a57e 100644
--- a/sound/stm/Makefile
+++ b/sound/stm/Makefile
@@ -9,6 +9,9 @@ snd-stm-stx710x-y := stx710x.o
 obj-$(CONFIG_SND_STM_STX7111) += snd-stm-stx7111.o
 snd-stm-stx7111-y := stx7111.o
 
+obj-$(CONFIG_SND_STM_STX7141) += snd-stm-stx7141.o
+snd-stm-stx7141-y := stx7141.o
+
 obj-$(CONFIG_SND_STM_STX7200) += snd-stm-stx7200.o
 snd-stm-stx7200-y := stx7200.o
 
diff --git a/sound/stm/reg_7141_audcfg.h b/sound/stm/reg_7141_audcfg.h
new file mode 100644
index 0000000..661ac3a
--- /dev/null
+++ b/sound/stm/reg_7141_audcfg.h
@@ -0,0 +1,160 @@
+#ifndef __SND_STM_7141_AUDCFG_H
+#define __SND_STM_7141_AUDCFG_H
+
+
+
+/*
+ * 7141_AUDCFG_IO_CTRL
+ */
+
+#define offset__7141_AUDCFG_IO_CTRL(ip) 0x00
+#define get__7141_AUDCFG_IO_CTRL(ip) readl(ip->base + \
+	offset__7141_AUDCFG_IO_CTRL(ip))
+#define set__7141_AUDCFG_IO_CTRL(ip, value) writel((value), ip->base + \
+	offset__7141_AUDCFG_IO_CTRL(ip))
+
+/* PCM_CLK_EN */
+
+#define shift__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip) 0
+#define mask__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip) 0x1
+#define get__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip) ((readl(ip->base + \
+	offset__7141_AUDCFG_IO_CTRL(ip)) >> \
+	shift__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip)) & \
+	mask__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip))
+#define set__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip, value) \
+	writel((readl(ip->base + offset__7141_AUDCFG_IO_CTRL(ip)) & \
+	~(mask__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip))) | (((value) & \
+	mask__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip)) << \
+	shift__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip)), ip->base + \
+	offset__7141_AUDCFG_IO_CTRL(ip))
+
+#define value__7141_AUDCFG_IO_CTRL__PCM_CLK_EN__INPUT(ip) 0x0
+#define mask__7141_AUDCFG_IO_CTRL__PCM_CLK_EN__INPUT(ip) \
+	(value__7141_AUDCFG_IO_CTRL__PCM_CLK_EN__INPUT(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip))
+#define set__7141_AUDCFG_IO_CTRL__PCM_CLK_EN__INPUT(ip) \
+	set__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip, \
+	value__7141_AUDCFG_IO_CTRL__PCM_CLK_EN__INPUT(ip))
+
+#define value__7141_AUDCFG_IO_CTRL__PCM_CLK_EN__OUTPUT(ip) 0x1
+#define mask__7141_AUDCFG_IO_CTRL__PCM_CLK_EN__OUTPUT(ip) \
+	(value__7141_AUDCFG_IO_CTRL__PCM_CLK_EN__OUTPUT(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip))
+#define set__7141_AUDCFG_IO_CTRL__PCM_CLK_EN__OUTPUT(ip) \
+	set__7141_AUDCFG_IO_CTRL__PCM_CLK_EN(ip, \
+	value__7141_AUDCFG_IO_CTRL__PCM_CLK_EN__OUTPUT(ip))
+
+/* SPDIFHDMI_EN */
+
+#define shift__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip) 3
+#define mask__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip) 0x1
+#define get__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip) ((readl(ip->base + \
+	offset__7141_AUDCFG_IO_CTRL(ip)) >> \
+	shift__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip)) & \
+	mask__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip))
+#define set__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip, value) \
+	writel((readl(ip->base + offset__7141_AUDCFG_IO_CTRL(ip)) & \
+	~(mask__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip))) | (((value) & \
+	mask__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip)) << \
+	shift__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip)), ip->base + \
+	offset__7141_AUDCFG_IO_CTRL(ip))
+
+#define value__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN__INPUT(ip) 0x0
+#define mask__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN__INPUT(ip) \
+	(value__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN__INPUT(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip))
+#define set__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN__INPUT(ip) \
+	set__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip, \
+	value__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN__INPUT(ip))
+
+#define value__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN__OUTPUT(ip) 0x1
+#define mask__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN__OUTPUT(ip) \
+	(value__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN__OUTPUT(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip))
+#define set__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN__OUTPUT(ip) \
+	set__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN(ip, \
+	value__7141_AUDCFG_IO_CTRL__SPDIFHDMI_EN__OUTPUT(ip))
+
+/* PCMPLHDMI_EN */
+
+#define shift__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip) 5
+#define mask__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip) 0x1
+#define get__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip) ((readl(ip->base + \
+	offset__7141_AUDCFG_IO_CTRL(ip)) >> \
+	shift__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip)) & \
+	mask__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip))
+#define set__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip, value) \
+	writel((readl(ip->base + offset__7141_AUDCFG_IO_CTRL(ip)) & \
+	~(mask__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip))) | (((value) & \
+	mask__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip)) << \
+	shift__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip)), ip->base + \
+	offset__7141_AUDCFG_IO_CTRL(ip))
+
+#define value__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN__INPUT(ip) 0x0
+#define mask__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN__INPUT(ip) \
+	(value__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN__INPUT(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip))
+#define set__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN__INPUT(ip) \
+	set__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip, \
+	value__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN__INPUT(ip))
+
+#define value__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN__OUTPUT(ip) 0x1
+#define mask__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN__OUTPUT(ip) \
+	(value__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN__OUTPUT(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip))
+#define set__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN__OUTPUT(ip) \
+	set__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN(ip, \
+	value__7141_AUDCFG_IO_CTRL__PCMPLHDMI_EN__OUTPUT(ip))
+
+/* CLKREC_SEL */
+
+#define shift__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip) 9
+#define mask__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip) 0x3
+#define get__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip) ((readl(ip->base + \
+	offset__7141_AUDCFG_IO_CTRL(ip)) >> \
+	shift__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip)) & \
+	mask__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip))
+#define set__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip, value) \
+	writel((readl(ip->base + offset__7141_AUDCFG_IO_CTRL(ip)) & \
+	~(mask__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip))) | (((value) & \
+	mask__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip)) << \
+	shift__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip)), ip->base + \
+	offset__7141_AUDCFG_IO_CTRL(ip))
+
+#define value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPLHDMI(ip) 0x0
+#define mask__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPLHDMI(ip) \
+	(value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPLHDMI(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip))
+#define set__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPLHDMI(ip) \
+	set__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip, \
+	value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPLHDMI(ip))
+
+#define value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__SPDIFHDMI(ip) 0x1
+#define mask__7141_AUDCFG_IO_CTRL__CLKREC_SEL__SPDIFHDMI(ip) \
+	(value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__SPDIFHDMI(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip))
+#define set__7141_AUDCFG_IO_CTRL__CLKREC_SEL__SPDIFHDMI(ip) \
+	set__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip, \
+	value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__SPDIFHDMI(ip))
+
+#define value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPL1(ip) 0x2
+#define mask__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPL1(ip) \
+	(value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPL1(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip))
+#define set__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPL1(ip) \
+	set__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip, \
+	value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPL1(ip))
+
+#define value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPL0(ip) 0x3
+#define mask__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPL0(ip) \
+	(value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPL0(ip) << \
+	shift__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip))
+#define set__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPL0(ip) \
+	set__7141_AUDCFG_IO_CTRL__CLKREC_SEL(ip, \
+	value__7141_AUDCFG_IO_CTRL__CLKREC_SEL__PCMPL0(ip))
+
+#endif
diff --git a/sound/stm/stx7141.c b/sound/stm/stx7141.c
new file mode 100644
index 0000000..347eb18
--- /dev/null
+++ b/sound/stm/stx7141.c
@@ -0,0 +1,597 @@
+/*
+ *   STMicrolectronics STx7141 SoC description & audio glue driver
+ *
+ *   Copyright (c) 2005-2007 STMicroelectronics Limited
+ *
+ *   Author: Stephen Gallimore <stephen.gallimore@st.com>
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program; if not, write to the Free Software
+ *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
+ *
+ */
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/irq.h>
+#include <asm/irq-ilc.h>
+#include <sound/driver.h>
+#include <sound/core.h>
+
+#define COMPONENT stx7141
+#include "common.h"
+#include "reg_7141_audcfg.h"
+
+
+
+/*
+ * ALSA module parameters
+ */
+
+static int index = -1; /* First available index */
+static char *id = "STx7141"; /* Default card ID */
+
+module_param(index, int, 0444);
+MODULE_PARM_DESC(index, "Index value for STx7141 audio subsystem card.");
+module_param(id, charp, 0444);
+MODULE_PARM_DESC(id, "ID string for STx7141 audio subsystem card.");
+
+
+
+/*
+ * Audio subsystem components & platform devices
+ */
+
+/* STx7141 audio glue */
+
+static struct platform_device stx7141_glue = {
+	.name          = "snd_stx7141_glue",
+	.id            = -1,
+	.num_resources = 1,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfe210200,
+			.end   = 0xfe21020b,
+		},
+	}
+};
+
+/* Frequency synthesizers */
+
+static struct platform_device fsynth = {
+	.name          = "snd_fsynth",
+	.id            = -1,
+	.num_resources = 1,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfe210000,
+			.end   = 0xfe21004f,
+		},
+	},
+	.dev.platform_data = &(struct snd_stm_fsynth_info) {
+		.ver = 4,
+		.channels_from = 0,
+		.channels_to = 3,
+	},
+};
+
+/* Internal DACs */
+
+static struct platform_device conv_int_dac = {
+	.name          = "snd_conv_int_dac",
+	.id            = -1,
+	.num_resources = 1,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfe210100,
+			.end   = 0xfe210103,
+		},
+	},
+	.dev.platform_data = &(struct snd_stm_conv_int_dac_info) {
+		.ver = 4,
+		.source_bus_id = "snd_pcm_player.1",
+		.channel_from = 0,
+		.channel_to = 1,
+	},
+};
+
+/* PCM players  */
+
+static struct platform_device pcm_player_0 = {
+	.name          = "snd_pcm_player",
+	.id            = 0,
+	.num_resources = 2,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfd101000,
+			.end   = 0xfd101027,
+		},
+		{
+			.flags = IORESOURCE_IRQ,
+			.start = ILC_IRQ(101),
+			.end   = ILC_IRQ(101),
+		},
+	},
+	.dev.platform_data = &(struct snd_stm_pcm_player_info) {
+		.name = "PCM player #0",
+		.ver = 6,
+		.card_device = 0,
+		.fsynth_bus_id = "snd_fsynth",
+		.fsynth_output = 3,
+		.channels = 10,
+		.fdma_initiator = 0,
+		.fdma_request_line = 39, /* TODO: CHECK THIS */
+	},
+};
+
+static struct platform_device pcm_player_1 = {
+	.name          = "snd_pcm_player",
+	.id            = 1,
+	.num_resources = 2,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfd101800,
+			.end   = 0xfd101827,
+		},
+		{
+			.flags = IORESOURCE_IRQ,
+			.start = ILC_IRQ(102),
+			.end   = ILC_IRQ(102),
+		},
+	},
+	.dev.platform_data = &(struct snd_stm_pcm_player_info) {
+		.name = "PCM player #1",
+		.ver = 6,
+		.card_device = 1,
+		.fsynth_bus_id = "snd_fsynth",
+		.fsynth_output = 1,
+		.channels = 2,
+		.fdma_initiator = 0,
+		.fdma_request_line = 40,
+	},
+};
+
+static struct platform_device pcm_player_2 = {
+	.name          = "snd_pcm_player",
+	.id            = 2,
+	.num_resources = 2,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfd104d00,
+			.end   = 0xfd104d27,
+		},
+		{
+			.flags = IORESOURCE_IRQ,
+			.start = ILC_IRQ(137),
+			.end   = ILC_IRQ(137),
+		},
+	},
+	.dev.platform_data = &(struct snd_stm_pcm_player_info) {
+		.name = "PCM player HDMI",
+		.ver = 6,
+		.card_device = 2,
+		.fsynth_bus_id = "snd_fsynth",
+		.fsynth_output = 0,
+		.channels = 8,
+		.fdma_initiator = 0,
+		.fdma_request_line = 47,
+	},
+};
+
+
+/*
+ * SPDIF player
+ */
+
+static struct platform_device spdif_player = {
+	.name          = "snd_spdif_player",
+	.id            = -1,
+	.num_resources = 2,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfd104c00,
+			.end   = 0xfd104c43,
+		},
+		{
+			.flags = IORESOURCE_IRQ,
+			.start = ILC_IRQ(136),
+			.end   = ILC_IRQ(136),
+		},
+	},
+	.dev.platform_data = &(struct snd_stm_spdif_player_info) {
+		.name = "SPDIF player (HDMI)",
+		.ver = 4,
+		.card_device = 3,
+		.fsynth_bus_id = "snd_fsynth",
+		.fsynth_output = 2,
+		.fdma_initiator = 0,
+		.fdma_request_line = 48,
+	},
+};
+
+/* I2S to SPDIF converters */
+
+static struct platform_device conv_i2sspdif_0 = {
+	.name          = "snd_conv_i2sspdif",
+	.id            = 0,
+	.num_resources = 2,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfd105000,
+			.end   = 0xfd105223,
+		},
+		{
+			.flags = IORESOURCE_IRQ,
+			.start = ILC_IRQ(135),
+			.end   = ILC_IRQ(135),
+		}
+	},
+	.dev.platform_data = &(struct snd_stm_conv_i2sspdif_info) {
+		.ver = 4,
+		.source_bus_id = "snd_pcm_player.2",
+		.channel_from = 0,
+		.channel_to = 1,
+	},
+};
+
+static struct platform_device conv_i2sspdif_1 = {
+	.name          = "snd_conv_i2sspdif",
+	.id            = 1,
+	.num_resources = 2,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfd105400,
+			.end   = 0xfd105623,
+		},
+		{
+			.flags = IORESOURCE_IRQ,
+			.start = ILC_IRQ(134),
+			.end   = ILC_IRQ(134),
+		}
+	},
+	.dev.platform_data = &(struct snd_stm_conv_i2sspdif_info) {
+		.ver = 4,
+		.source_bus_id = "snd_pcm_player.2",
+		.channel_from = 2,
+		.channel_to = 3,
+	},
+};
+
+static struct platform_device conv_i2sspdif_2 = {
+	.name          = "snd_conv_i2sspdif",
+	.id            = 2,
+	.num_resources = 2,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfd105800,
+			.end   = 0xfd105a23,
+		},
+		{
+			.flags = IORESOURCE_IRQ,
+			.start = ILC_IRQ(133),
+			.end   = ILC_IRQ(133),
+		}
+	},
+	.dev.platform_data = &(struct snd_stm_conv_i2sspdif_info) {
+		.ver = 4,
+		.source_bus_id = "snd_pcm_player.2",
+		.channel_from = 4,
+		.channel_to = 5,
+	},
+};
+
+static struct platform_device conv_i2sspdif_3 = {
+	.name          = "snd_conv_i2sspdif",
+	.id            = 3,
+	.num_resources = 2,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfd105c00,
+			.end   = 0xfd105e23,
+		},
+		{
+			.flags = IORESOURCE_IRQ,
+			.start = ILC_IRQ(132),
+			.end   = ILC_IRQ(132),
+		}
+	},
+	.dev.platform_data = &(struct snd_stm_conv_i2sspdif_info) {
+		.ver = 4,
+		.source_bus_id = "snd_pcm_player.2",
+		.channel_from = 6,
+		.channel_to = 7,
+	},
+};
+
+/* PCM reader */
+
+static struct platform_device pcm_reader_0 = {
+	.name          = "snd_pcm_reader",
+	.id            = 0,
+	.num_resources = 2,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfd102000,
+			.end   = 0xfd102027,
+		},
+		{
+			.flags = IORESOURCE_IRQ,
+			.start = ILC_IRQ(103),
+			.end   = ILC_IRQ(103),
+		},
+	},
+	.dev.platform_data = &(struct snd_stm_pcm_reader_info) {
+		.name = "PCM Reader #0",
+		.ver = 4,
+		.card_device = 4,
+		.channels = 2,
+		.fdma_initiator = 0,
+		.fdma_request_line = 41,
+	},
+};
+
+static struct platform_device pcm_reader_1 = {
+	.name          = "snd_pcm_reader",
+	.id            = 1,
+	.num_resources = 2,
+	.resource      = (struct resource []) {
+		{
+			.flags = IORESOURCE_MEM,
+			.start = 0xfd103000,
+			.end   = 0xfd103027,
+		},
+		{
+			.flags = IORESOURCE_IRQ,
+			.start = ILC_IRQ(104),
+			.end   = ILC_IRQ(104),
+		},
+	},
+	.dev.platform_data = &(struct snd_stm_pcm_reader_info) {
+		.name = "PCM Reader #1",
+		.ver = 4,
+		.card_device = 5,
+		.channels = 2,
+		.fdma_initiator = 0,
+		.fdma_request_line = 42,
+	},
+};
+
+static struct platform_device *snd_stm_stx7141_devices[] = {
+	&stx7141_glue,
+	&fsynth,
+	&conv_int_dac,
+	&pcm_player_0,
+	&pcm_player_1,
+	&pcm_player_2,
+	&spdif_player,
+	&conv_i2sspdif_0,
+	&conv_i2sspdif_1,
+	&conv_i2sspdif_2,
+	&conv_i2sspdif_3,
+	&pcm_reader_0,
+	&pcm_reader_1,
+};
+
+
+
+/*
+ * Audio glue driver implementation
+ */
+
+struct snd_stm_stx7141_glue {
+	int ver;
+
+	struct resource *mem_region;
+	void *base;
+
+	struct snd_info_entry *proc_entry;
+
+	snd_stm_magic_field;
+};
+
+static void snd_stm_stx7141_glue_dump_registers(struct snd_info_entry *entry,
+		struct snd_info_buffer *buffer)
+{
+	struct snd_stm_stx7141_glue *stx7141_glue = entry->private_data;
+
+	snd_stm_assert(stx7141_glue, return);
+	snd_stm_magic_assert(stx7141_glue, return);
+
+	snd_iprintf(buffer, "--- snd_stx7141_glue ---\n");
+	snd_iprintf(buffer, "base = 0x%p\n", stx7141_glue->base);
+
+	snd_iprintf(buffer, "AUDCFG_IO_CTRL (offset 0x00) = 0x%08x\n",
+			get__7141_AUDCFG_IO_CTRL(stx7141_glue));
+
+	snd_iprintf(buffer, "\n");
+}
+
+static int __init snd_stm_stx7141_glue_probe(struct platform_device *pdev)
+{
+	int result = 0;
+	struct snd_stm_stx7141_glue *stx7141_glue;
+
+	snd_stm_printd(0, "--- Probing device '%s'...\n", pdev->dev.bus_id);
+
+	stx7141_glue = kzalloc(sizeof(*stx7141_glue), GFP_KERNEL);
+	if (!stx7141_glue) {
+		snd_stm_printe("Can't allocate memory "
+				"for a device description!\n");
+		result = -ENOMEM;
+		goto error_alloc;
+	}
+	snd_stm_magic_set(stx7141_glue);
+
+	result = snd_stm_memory_request(pdev, &stx7141_glue->mem_region,
+			&stx7141_glue->base);
+	if (result < 0) {
+		snd_stm_printe("Memory region request failed!\n");
+		goto error_memory_request;
+	}
+
+	/* Additional procfs info */
+
+	snd_stm_info_register(&stx7141_glue->proc_entry, "stx7141_glue",
+			snd_stm_stx7141_glue_dump_registers, stx7141_glue);
+
+	/* Done now */
+
+	platform_set_drvdata(pdev, stx7141_glue);
+
+	snd_stm_printd(0, "--- Probed successfully!\n");
+
+	return result;
+
+error_memory_request:
+	snd_stm_magic_clear(stx7141_glue);
+	kfree(stx7141_glue);
+error_alloc:
+	return result;
+}
+
+static int __exit snd_stm_stx7141_glue_remove(struct platform_device *pdev)
+{
+	struct snd_stm_stx7141_glue *stx7141_glue =
+			platform_get_drvdata(pdev);
+
+	snd_stm_assert(stx7141_glue, return -EINVAL);
+	snd_stm_magic_assert(stx7141_glue, return -EINVAL);
+
+	snd_stm_info_unregister(stx7141_glue->proc_entry);
+
+	snd_stm_memory_release(stx7141_glue->mem_region, stx7141_glue->base);
+
+	snd_stm_magic_clear(stx7141_glue);
+	kfree(stx7141_glue);
+
+	return 0;
+}
+
+static struct platform_driver snd_stm_stx7141_glue_driver = {
+	.driver = {
+		.name = "snd_stx7141_glue",
+	},
+	.probe = snd_stm_stx7141_glue_probe,
+	.remove = snd_stm_stx7141_glue_remove,
+};
+
+
+
+/*
+ * Audio initialization
+ */
+
+static int __init snd_stm_stx7141_init(void)
+{
+	int result;
+	struct snd_card *card;
+
+	snd_stm_printd(0, "snd_stm_stx7141_init()\n");
+
+	if (cpu_data->type != CPU_STX7141) {
+		snd_stm_printe("Not supported (other than STx7141) SOC "
+				"detected!\n");
+		result = -EINVAL;
+		goto error_soc_type;
+	}
+
+	/* Cut 2.0 presumably will bring something new into the
+	 * matter, so above configuration must be checked!
+	 * - transfer_sizes (FIFO sizes has changed) */
+	WARN_ON(cpu_data->cut_major > 1);
+
+	/* Ugly but quick hack to have SPDIF player & I2S to SPDIF
+	 * converters enabled without loading STMFB...
+	 * TODO: do this in some sane way! */
+	{
+		void *hdmi_gpout = ioremap(0xfd104020, 4);
+		writel(readl(hdmi_gpout) | 0x3, hdmi_gpout);
+		iounmap(hdmi_gpout);
+	}
+
+	result = platform_driver_register(&snd_stm_stx7141_glue_driver);
+	result = 0;
+	if (result != 0) {
+		snd_stm_printe("Failed to register audio glue driver!\n");
+		goto error_driver_register;
+	}
+
+	card = snd_stm_card_new(index, id, THIS_MODULE);
+	if (card == NULL) {
+		snd_stm_printe("ALSA card creation failed!\n");
+		result = -ENOMEM;
+		goto error_card_new;
+	}
+	strcpy(card->driver, "STx7141");
+	strcpy(card->shortname, "STx7141 audio subsystem");
+	snprintf(card->longname, 79, "STMicroelectronics STx7141 cut %d "
+			"SOC audio subsystem", cpu_data->cut_major);
+
+	result = snd_stm_add_platform_devices(snd_stm_stx7141_devices,
+			ARRAY_SIZE(snd_stm_stx7141_devices));
+	if (result != 0) {
+		snd_stm_printe("Failed to add platform devices!\n");
+		goto error_add_devices;
+	}
+
+	result = snd_stm_card_register();
+	if (result != 0) {
+		snd_stm_printe("Failed to register ALSA cards!\n");
+		goto error_card_register;
+	}
+
+	return 0;
+
+error_card_register:
+	snd_stm_remove_platform_devices(snd_stm_stx7141_devices,
+			ARRAY_SIZE(snd_stm_stx7141_devices));
+error_add_devices:
+	snd_stm_card_free();
+error_card_new:
+	platform_driver_unregister(&snd_stm_stx7141_glue_driver);
+error_driver_register:
+error_soc_type:
+	return result;
+}
+
+static void __exit snd_stm_stx7141_exit(void)
+{
+	snd_stm_printd(0, "snd_stm_stx7141_exit()\n");
+
+	snd_stm_card_free();
+
+	snd_stm_remove_platform_devices(snd_stm_stx7141_devices,
+			ARRAY_SIZE(snd_stm_stx7141_devices));
+
+	platform_driver_unregister(&snd_stm_stx7141_glue_driver);
+}
+
+MODULE_AUTHOR("Stephen Gallimore <stephen.gallimore@st.com>");
+MODULE_DESCRIPTION("STMicroelectronics STx7141 audio driver");
+MODULE_LICENSE("GPL");
+
+module_init(snd_stm_stx7141_init);
+module_exit(snd_stm_stx7141_exit);
